envoy/tracing/trace_driver.h

```diff
@@ -14,6 +14,45 @@ namespace Tracing {
 class Span;
 using SpanPtr = std::unique_ptr<Span>;
 
+/**
+ * The upstream sevice type.
+ */
+enum class ServiceType {
+  // Service type is unknown.
+  Unknown,
+  // Service is treated as HTTP.
+  Http,
+  // Service is treated as GoogleGrpc.
+  GoogleGrpc,
+  // Service is treated as EnvoyGrpc.
+  EnvoyGrpc
+};
+
+/**
+ * Contains upstream context information essential for the injectContext process.
+ *
+ * @param host Optional reference to the upstream host description.
+ * @param cluster_info Optional reference to the upstream cluster information.
+ * @param service_type The type of service the upstream context relates to.
+ * @param is_from_async_client Indicates if the injectContext originates from an asynchronous
+ * client.
+ */
+struct UpstreamContext {
+  UpstreamContext(const Upstream::HostDescriptionConstSharedPtr& host = nullptr,
+                  const Upstream::ClusterInfoConstSharedPtr& cluster_info = nullptr,
+                  const Tracing::ServiceType service_type = Tracing::ServiceType::Unknown,
+                  const bool is_from_async_client_ = false)
+      : host_(nullptr == host ? absl::nullopt_t : OptRef<const Upstream::HostDescription>(*host)),
+        cluster_info_(nullptr == cluster_info ? absl::nullopt_t
+                                              : OptRef<const Upstream::ClusterInfo>(*cluster_info)),
+        service_type_(service_type), is_from_async_client_(is_from_async_client_) {}
```

Comment:
Seems that you still kept the `cluster_info`. Is there any actually scenario that need it? (If you prefer to keep it, I am also OK.)

The constructor could be simpler. (And the `async_client_span` maybe more clear than the `is_from_async_client`?)

```
  UpstreamContext(const Upstream::HostDescription* host = nullptr,
                  const Upstream::ClusterInfo* cluster = nullptr,
                  const ServiceType service_type = ServiceType::Unknown,
                  const bool async_client_span = false)
      : host_(makeOptRefFromPtr(host)), cluster_(makeOptRefFromPtr(cluster)),
        service_type_(service_type), async_client_span_(async_client_span) {}
```

Reply:
Thank you for the quick review & suggestion! 

sry, meant to reply when addressing comments. Yeah, I prefer to remain cluster info for async client use-case, since some info can be used from the cluster.